core/templates/pages/exploration-player-page/services/question-player-engine.service.spec.ts

```diff
@@ -0,0 +1,873 @@
+    expect(languageCode).toBe('ab');
+  });
+
+  it('should always return false when calling \'isInPreviewMode()\'', () => {
+    let previewMode = questionPlayerEngineService.isInPreviewMode();
+
+    expect(previewMode).toBe(false);
+  });
+
+  it('should show warning message while loading a question ' +
+    'if the question html content is null', () => {
+    let successHandler = jasmine.createSpy('success');
+    let failHandler = jasmine.createSpy('fail');
+
+    singleQuestionBackendDict.question_state_data
+      .content.html = null;
+    let alertsServiceSpy =
+      spyOn(alertsService, 'addWarning').and.returnValue();
+    spyOn(expressionInterpolationService, 'processHtml')
+      .and.callFake((html, envs) => html);
+
+    questionPlayerEngineService.init(
+      [singleQuestionBackendDict], successHandler, failHandler);
+
+    expect(alertsServiceSpy).toHaveBeenCalledWith('Expression parsing error.');
+  });
+
+  it('should not load questions if there are no questions', () => {
+    let successHandler = jasmine.createSpy('success');
+    let failHandler = jasmine.createSpy('fail');
+
+    questionPlayerEngineService.init(
+      [], successHandler, failHandler);
+    expect(successHandler).not.toHaveBeenCalled();
+    expect(failHandler).toHaveBeenCalled();
+  });
+
+  describe('on submitting answer ', () => {
+    it('should submit answer correctly', () => {
+      let successCallback = jasmine.createSpy('success');
+      let successHandler = jasmine.createSpy('success');
+      let failHandler = jasmine.createSpy('fail');
+      let answer = 'answer';
+      let interactionRulesService: InteractionRulesService;
+      let answerClassificationResult = new AnswerClassificationResult(
+        outcomeObjectFactory
+          .createNew('default', '', '', []), 1, 0, 'default_outcome'
+      );
+      answerClassificationResult.outcome.labelledAsCorrect = true;
+
+
+      spyOn(answerClassificationService, 'getMatchingClassificationResult')
+        .and.returnValue(answerClassificationResult);
+
+      questionPlayerEngineService.init(
+        multipleQuestionsBackendDict, successHandler, failHandler);
+      questionPlayerEngineService.submitAnswer(
+        answer, interactionRulesService, successCallback);
+
+      expect(questionPlayerEngineService.isAnswerBeingProcessed()).toBe(false);
+      expect(successCallback).toHaveBeenCalled();
+    });
+
+    it('should not submit answer again if the answer ' +
+      'is already being processed', () => {
+      let successCallback = jasmine.createSpy('success');
+      let answer = 'answer';
+      let interactionRulesService: InteractionRulesService;
+
+
+      spyOn(answerClassificationService, 'getMatchingClassificationResult');
+
+      questionPlayerEngineService.setAnswerIsBeingProcessed(true);
+      questionPlayerEngineService.submitAnswer(
+        answer, interactionRulesService, successCallback);
+
+      expect(successCallback).not.toHaveBeenCalled();
```

Comment:
For this to be robust you would need to also demonstrate that the successCallback is called when the answer is not already being processed.

Reply:
There is already a check related to that, but with a different description `it('should submit answer correctly'`,
do we need to add an extra case for that?

Reply:
Can we change that to `should call success callback if the submitted answer is correct`?